services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/service_liveness.py

```diff
@@ -0,0 +1,104 @@
+from tenacity.stop import stop_after_delay
+from tenacity.wait import wait_fixed
+
+_logger = logging.getLogger(__name__)
+
+
+_DEFAULT_CHECK_INTERVAL: Final[timedelta] = timedelta(seconds=1)
+_DEFAULT_TIMEOUT_INTERVAL: Final[timedelta] = timedelta(seconds=30)
+
+
+class CouldNotReachServiceError(PydanticErrorMixin, Exception):
+    msg_template: str = "Could not contact service '{service_name}' at '{endpoint}'. Look above for details."
+
+
+def _before_sleep_log(
+    logger: logging.Logger, service_name: str, endpoint: str
+) -> Callable[[RetryCallState], None]:
+    def log_it(retry_state: RetryCallState) -> None:
+        assert retry_state  # nosec
+        assert retry_state.next_action  # nosec
+        logger.warning(
+            "Retrying to contact '%s' at '%s' in %s seconds.",
+            service_name,
+            endpoint,
+            retry_state.next_action.sleep,
+        )
+
+    return log_it
+
+
+async def _attempt_to_wait_for_handler(
+    async_handler: Callable[..., Awaitable],
+    *args,
+    service_name: str,
+    endpoint: str,
+    check_interval: timedelta,
+    timeout: timedelta,
+    **kwargs,
+) -> None:
+    async for attempt in AsyncRetrying(
+        wait=wait_fixed(check_interval),
+        stop=stop_after_delay(timeout.total_seconds()),
+        before_sleep=_before_sleep_log(_logger, service_name, endpoint),
+        reraise=True,
+    ):
+        with attempt:
+            if await async_handler(*args, **kwargs) is False:
+                raise TryAgain
+
+
+async def wait_for_service_liveness(
+    async_handler: Callable[..., Awaitable],
+    *args,
+    service_name: str,
+    endpoint: str,
+    check_interval: timedelta = _DEFAULT_CHECK_INTERVAL,
+    timeout: timedelta = _DEFAULT_TIMEOUT_INTERVAL,
+    **kwargs,
+) -> None:
+    """waits for async_handler to return ``True`` or ``None`` instead of
+    raising errors or returning ``False``
+
+    Arguments:
+        async_handler -- handler to execute
+        service_name -- service reference for whom investigates the logs
+        endpoint -- endpoint address for whom investigates the logs
+
+    Keyword Arguments:
+        check_interval -- interval at which service check is ran (default: {_DEFAULT_CHECK_INTERVAL})
+        timeout -- stops trying to contact service and raises ``CouldNotReachServiceError``
+            (default: {_DEFAULT_TIMEOUT_INTERVAL})
+
+    Raises:
+        CouldNotReachServiceError: if it was not able to contact the service in time
+    """
+    #
+
+    try:
+        start = time.time()
+        await _attempt_to_wait_for_handler(
+            async_handler,
+            *args,
+            service_name=service_name,
+            endpoint=endpoint,
+            check_interval=check_interval,
+            timeout=timeout,
+            **kwargs,
+        )
+        elapsed = (time.time() - start) * 1000
+        _logger.info(
+            "Service '%s' found at '%s' after %.2f ms", service_name, endpoint, elapsed
```

Comment:
I think all that timing stuff is given by tenacity in the retry_state.statistics or so. please search in the code for similar things